### ISSUE TITLE
Ethos on X20s is painfully slow, when LUA sources are used

### ISSUE THREAD

**Issue by azaz44** [FrSkyRC/ETHOS-Feedback-Community#2976]
My radio runs painfully slow to the point where making changes to the model is really really uncomfortable. Just scrolling a list of logic switches or mixers is a pain, not to mention things like reordering items. My models do use a number of variables (var mixers) and logic switches, but nothing close to limits.

I tried to research this today and I think there's some problem with LUA sources.

Video: https://youtu.be/5l0N2N9kpsE

The first part is where radio is very slow. It's a clone of one of my real models. But I removed all LUA sources I normally use, and added just one LUA source which makes nothing. Here's the code:

```lua
local function sourceInit(source)
end

local function sourceWakeup(source)
end

local function init()
  system.registerSource({
    key = "aaa1",
    name = "aaa",
    init = sourceInit,
    wakeup = sourceWakeup
  })
end

return { init = init }
```

As you can see in video, radio runs very slow. Getting to the end of LS or mixers list takes a long time. And model selection screen loads model images so slow, that it isn't a useful feature at all. At the same time I spin the wheel as fast as I can.

At 1:40 I disable the LUA source (the one which does nothing).

And as you can see later, performance increase is by orders of magnitude.

I'm not sure where the problem is. If I create a new empty model, and use this LUA code, it works OK. It looks like a combination of multiple logic switches and mixers and a LUA source, even a single one which does nothing. Enabling more LUA scripts doesn't change much.

I suspect there's some costly processing which happens every frame, when at least one LUA source is active, and which involves all sources in the model. So if we have more sources, and at least one LUA, we have a problem.

Running X20S, firmware 1.4.11.
I'm sorry for not testing in nightlies, but I need to run production software now every day.

**Comment by strgaltdel** [comment]
hi azaz44
sounds similar to my experience
maybe some infos from Bertrand could be helpful for you?
(my failure was declaring and reading value in every handler loop, but this seems not the case in this issue)
he mentioned at least number of LSWs and SFs shouldnt hurt so much
#2830

**Comment by azaz44** [comment]
It might be related, but hm, in my case it is any script causing issues. Even empty. And the performance penalty is really huge.

**Comment by bsongis-frsky** [comment]
Such a source caused the main task check for model source changes in each loop. Will be fixed in 1.4.13

**Comment by azaz44** [comment]
> Such a source caused the main task check for model source changes in each loop. Will be fixed in 1.4.13

I wonder if this is related to changes made as a solution of this issue: https://github.com/FrSkyRC/ETHOS-Feedback-Community/issues/2316

I can't really find description there on how it is implemented. But IIRC (maybe it was discussed on Discord), after changes, Ethos is notifying / changing `source` objects in LUA scripts to keep them up to date. So for example, if user reorders logic switches, and LSW2 becomes LSW1, LUA `source` object will be updated.

It sounds to me that this might be relates, so letting you know, so the other thing doesn't get broken by accident.

**Comment by bsongis-frsky** [comment]
Exactly, this is when the problem has been introduced.
